src/parser.ts

```diff
@@ -3986,7 +3984,19 @@ export function parseMemberOrUpdateExpression(
         });
     }
 
-    return parseMemberOrUpdateExpression(parser, context, expr, 0, start, line, column);
+    expr = parseMemberOrUpdateExpression(parser, context, expr, 0, 1, start, line, column);
+  }
+
+  // Finalize ChainExpression
+  if (inChain === 0 && (parser.flags & Flags.HasOptionalChaining) === Flags.HasOptionalChaining) {
+    parser.flags = (parser.flags | Flags.HasOptionalChaining) ^ Flags.HasOptionalChaining;
+    // FIXME: following line got issue with a?.[++x]
+    // parser.assignable = AssignmentKind.CannotAssign;
```

Comment:
I got some issue around invalidating assignment.
With the above line, `a?.[++x]` will fail (but it should be valid).
With out the above line, `({ a: x?.obj['a'] } = {})` will pass (but it should be invalid).

Reply:
That's the meriyah code :)  What if you change to `parser.assignable = AssignmentKind.Assignable` and if that doesn\t work, just move the `parser.assignable = AssignmentKind.CannotAssign` to another location

Reply:
The `({ a: x?.obj['a'] } = {})` has something to do with DestructuringKind, I could not figure out.

Reply:
`({ a: x?.obj['a'] } = {})` should be invalid. If it doesn't faill. Just comment it out

Reply:
It should be invalid. The problem is I could not find way to invalidate it properly.

Reply:
Comment it out for now. It's a nightmare to fix destructing issues in Meriyah :)

Reply:
BTW. Why do you use a mutual parser flag for the chaining?

Reply:
I am not sure how to record the usage of optional in the chain, it's probably not needed.
I guess I can traverse the chain to find out "one or more `ChainElement` nodes that are `optional:true`" when finalising the chain here.

Reply:
I see. Estree complicated it with their implementation because of lack of parenthized expression.  And their `wavy` PR complicates this further. It should use the `ChainElement`